dvc/repo/metrics/show.py

```diff
@@ -67,6 +72,50 @@ def _read_metric(path, typ=None, xpath=None):
     return ret
 
 
+def _collect_metrics(self, path, recursive, typ, xpath):
+    outs = [out for stage in self.stages() for out in stage.outs]
+
+    if path:
+        try:
+            outs = self.find_outs_by_path(path, outs=outs, recursive=recursive)
+        except OutputNotFoundError:
+            return []
+
+    res = []
+    for o in outs:
+        if not o.metric:
+            continue
+
+        if not typ and isinstance(o.metric, dict):
+            t = o.metric.get(o.PARAM_METRIC_TYPE, typ)
+            x = o.metric.get(o.PARAM_METRIC_XPATH, xpath)
+        else:
+            t = typ
+            x = xpath
+
+        res.append((o, t, x))
+
+    return res
+
+
+def _read_metrics(self, metrics):
+    res = {}
+    for out, typ, xpath in metrics:
+        assert out.scheme == "local"
+        if out.use_cache:
+            path = self.cache.local.get(out.checksum)
```

Comment:
btw, what happens if file is not in cache? Like if just cloned a repo, did `dvc pull` and run `dvc metrics show -a`. Cache will be empty for other branches. Should we catch this and show something - you need to run `dvc pull -a`? Or maybe we even need something like `dvc pull -m -a` - pull only cached metrics.

Reply:
Hm, that is a great point! Yes, I'll add that. Thank you! :)

Reply:
Added https://github.com/iterative/dvc/issues/1609 for it. Will have to do something similar for `dvc commit`, so I'll get back to it later. For now we are not introducing any regressions in this.